### PR TITLE
[Blackjack - Step1] Submit Farhana's mission

### PR DESCRIPTION
**Pull request by farhanaahmed**
This PR implements the core logic of a simplified Blackjack game using the MVC architecture in Kotlin.
1. Is using an interface for Participant a good design decision here, or would an abstract class offer better code reuse in this case?
2. Does the Participants class help improve clarity and structure, or could it be simplified further?
3. Any suggestions for improving testability?

### CODE REVIEW

File: src/main/kotlin/blackjack/view/InputView.kt [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package blackjack.view
+
+class InputView {
```
Comment (pkeugine) [inline src/main/kotlin/blackjack/view/InputView.kt]:
```suggestion
object InputView {
```

We can make the view classes as objects.
What do you think are the differences between a class and an object?
Official Kotlin documentation can be a great source to find this out :)

File: src/main/kotlin/blackjack/view/InputView.kt [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+            throw IllegalArgumentException("Input can not be empty.")
+        }
+    }
+
+    fun askToDrawCard(name: String): Boolean {
+        println("Would $name like to draw another card? (y for yes, n for no)")
+        val choice = readln()
+        return when (choice) {
+            "y", "Y" -> true
+            "n", "N" -> false
```
Comment (pkeugine) [inline src/main/kotlin/blackjack/view/InputView.kt:24]: Nice job on considering upper case inputs as well 👍

File: README.md [hunk 3]
```diff
@@ -1 +1,48 @@
...
+- Input View 
+- [x] ask user inputs for player name
+- [x] get list of names from the user input
+
+- Card 
+- [x] Visibility of the card implemented by flipCard() 
+- [x] Determine the value of each card
+- [x] Create Enum class - Ranks
+- [x] Create Class - Card
+- [x] Create Enum class - Suits
```
Comment (pkeugine) [inline README.md]: Very clean README and commit logs 💯 
It was very easy to follow through the requirements and workflow of this PR.

File: src/main/kotlin/blackjack/model/Suit.kt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package blackjack.model
+
+enum class Suit(val symbol: Char) {
+    HEARTS('♥'),
+    DIAMONDS('♦'),
+    CLUBS('♣'),
+    SPADES('♠'),
```
Comment (pkeugine) [inline src/main/kotlin/blackjack/model/Suit.kt]: I see visual effects that makes the game even more fun 👍
But here's my thought:

Since the visual effects are used in console based output, I think they belong to the OutputView.
When using different output strategy other than console, we could use other visual component rather than a String. How does this sound?

File: src/main/kotlin/blackjack/model/Dealer.kt [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package blackjack.model
+
+class Dealer(
+    override val name: String = "Dealer",
+    override var isActive: Boolean = true,
+    override val cardsInHand: MutableList<Card> = mutableListOf(),
+) : Participant {
+    fun mustDraw(totalValueOfCards: Int) = totalValueOfCards <= 16
```
Comment (pkeugine) [inline src/main/kotlin/blackjack/model/Dealer.kt]: I think this 16 could have a name since it's a magic number.

File: src/test/kotlin/blackjack/model/PlayerTest.kt [hunk 6]
```diff
@@ -0,0 +1,45 @@
+package blackjack.model
+
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+
+class PlayerTest {
+    val player = Player("Alex")
```
Comment (pkeugine) [inline src/test/kotlin/blackjack/model/PlayerTest.kt]: It's interesting to see various tests use the same `player`.
The tests get passed for now, but this kind of test fixture setup is not recommended
since in certain cases, tests will share the same fixture.

When tests share the same fixture, all the tests that use the fixture will get have strong coupling,resulting in unexpected test results.

Why don't we make tests use their own `Player` in order to prevent these cases?
Also, I want you to tell me why the tests get passed currently, and when this might fail.

File: src/main/kotlin/blackjack/controller/Game.kt [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        }
+    }
+
+    fun askPlayersToDraw(players: List<Player>) {
+        players.forEach {
+            do {
+                val answer = inputView.askToDrawCard(it.name)
+                if (answer) {
+                    it.drawCard(deck.giveCard())
+                    it.updateActiveStatus(it.calculateTotalValueOfCards())
```
Comment (pkeugine) [inline src/main/kotlin/blackjack/controller/Game.kt]: Since these logics are all about `Player`, why don't we keep it simple and let `Player` do all these with one simple call. :)

File: src/main/kotlin/blackjack/model/Participant.kt [hunk 8]
```diff
@@ -0,0 +1,37 @@
+package blackjack.model
+
+interface Participant {
+    val name: String
+    var isActive: Boolean
+    val cardsInHand: MutableList<Card>
```
Comment (pkeugine) [inline src/main/kotlin/blackjack/model/Participant.kt]: Since `Participant` can evaluate total value of cards by itself,
I think `var isActive` is unnecessary since we can find out if they are active or not right away.
By having this `var isActive`, some other developer might change this by accident, creating a bug.
I'd like to know how you think about this :)

File: src/main/kotlin/blackjack/controller/Game.kt [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        }
+    }
+
+    fun compareFinalCards(players: List<Player>) {
+        players.forEach {
+            if (it.calculateTotalValueOfCards() < dealer.calculateTotalValueOfCards()) {
+                it.isActive = false
+            }
+        }
+    }
```
Comment (pkeugine) [inline src/main/kotlin/blackjack/controller/Game.kt]: This part can also be handled in `Player`.
By doing this, we'll be able to test this logic as well. 👍

File: src/main/kotlin/blackjack/model/Deck.kt [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+    fun getCards(): List<Card> {
+        return cards
+    }
+
+    fun giveCard(): Card {
+        require(cards.isNotEmpty()) { "Deck is empty!" }
+        return cards.removeAt(0)
+    }
+}
```
Comment (pkeugine) [inline src/main/kotlin/blackjack/model/Deck.kt]: Instead of giving a single card at a time, having a parameter that tells `Deck` the amount of cards needed would greatly improve this method's usage.

Nice validation on emptiness though! 👍

File: src/main/kotlin/blackjack/controller/Game.kt [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+        outputView.displayCardsWithTotalValue(dealer, players)
+        compareFinalCards(players)
+        outputView.displayFinalResults(dealer, players)
+    }
+
+    fun dealCardsToDealer() {
+        repeat(2) {
+            dealer.drawCard(deck.giveCard())
+        }
+    }
```
Comment (pkeugine) [inline src/main/kotlin/blackjack/controller/Game.kt]:
```suggestion
    private fun dealCardsToDealer() {
        repeat(2) {
            dealer.drawCard(deck.giveCard())
        }
    }
```

Let's keep things private if they aren't needed in outside world 🤫 .
Please check the whole project about this part.

File: src/main/kotlin/blackjack/view/OutputView.kt [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+        }
+    }
+
+    fun displayDealerInitialCards(dealer: Dealer) {
+        println("Dealer: ${dealer.cardsInHand[0].rank.face}${dealer.cardsInHand[0].suit.symbol}")
+    }
+
+    fun displayHandCards(player: Player) {
+        println(player.storePlayerHand())
+    }
```
Comment (pkeugine) [inline src/main/kotlin/blackjack/view/OutputView.kt]: I'd like to know what your personal (or team) convention is about the order of methods in a class.
Ordering with specific intention and keeping it across the whole project really helps the maintainability. If you haven't thought of one, it'll be nice to think about it in this step!

File: src/test/kotlin/blackjack/model/CardTest.kt [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+import org.junit.jupiter.api.Test
+
+class CardTest {
+    @Test
+    fun `should create a card with given rank and suit`() {
+        val card = Card(Rank.TWO, Suit.HEARTS)
+        assertThat(card.rank.toString()).isEqualTo("TWO")
+        assertThat(card.suit.toString()).isEqualTo("HEARTS")
+    }
+}
```
Comment (pkeugine) [inline src/test/kotlin/blackjack/model/CardTest.kt]:
```suggestion
    @Test
    fun `should create a card with given rank and suit`() {
        val card = Card(Rank.TWO, Suit.HEARTS)
        assertThat(card.rank).isEqualTo(Rank.TWO)
        assertThat(card.suit).isEqualTo(Suit.HEARTS)
    }
}
```

Although it might seem not important, I think it would be better to use the enum itself
when asserting the contents inside a card.

I recommend using `toString()` only for logging or debugging purposes,
rather than as making an output, as part of a logic, or as test validation value.

Well, I guess `TWO` and `HEARTS` could count as debugging / logging info as well. 😂

File: src/test/kotlin/blackjack/model/DeckTest.kt [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+
+class DeckTest {
+    val deck = Deck()
+
+    @Test
+    fun `deck should contain 52 cards`() {
+        assertThat(deck.getCards().count()).isEqualTo(52)
```
Comment (pkeugine) [inline src/test/kotlin/blackjack/model/DeckTest.kt]:
```suggestion
        assertThat(deck.count()).isEqualTo(52)
```

Why don't we let deck count the cards by itself? 
Let's try our best to take out the value inside of a class only for output purposes. :)

File: src/test/kotlin/blackjack/model/PlayerTest.kt [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    @Test
+    fun `should calculate sum of values of cards in hand`() {
+        val card1 = Card(Rank.TWO, Suit.HEARTS)
+        val card2 = Card(Rank.THREE, Suit.HEARTS)
+        player.drawCard(card1)
+        player.drawCard(card2)
+        assertThat(player.calculateTotalValueOfCards()).isEqualTo(5)
+    }
```
Comment (pkeugine) [inline src/test/kotlin/blackjack/model/PlayerTest.kt]: It would be great to cover more cases of different kinds of cards.
This will allow us to refactor with more confidence.

File: src/main/kotlin/blackjack/model/Participant.kt [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+
+    fun drawCard(card: Card) {
+        cardsInHand += card
+    }
+
+    private fun checkAces() = cardsInHand.count { card -> card.rank == Rank.ACE }
+
+    fun calculateTotalValueOfCards(): Int {
+        var totalValueOfCards = cardsInHand.sumOf { card -> card.rank.value }
+        var aceCounter = checkAces()
```
Comment (pkeugine) [inline src/main/kotlin/blackjack/model/Participant.kt]: Just out of curiosity,
how come `totalValueOfCards` logic wasn't extracted into it's own method
while `checkAces` got one?

File: src/main/kotlin/blackjack/view/InputView.kt [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+            else -> throw IllegalArgumentException("Input can only contain y for yes, n for no")
```
Comment (pkeugine) [inline src/main/kotlin/blackjack/view/InputView.kt]: Screenshot 2025-07-11 at 2 35 36 PM

It would be great to forgive an invalid input! 😂

File: src/main/kotlin/blackjack/model/Participant.kt [hunk 18]
```diff
@@ -0,0 +1,37 @@
+package blackjack.model
+
+interface Participant {
```
Comment (pkeugine) [inline src/main/kotlin/blackjack/model/Participant.kt]: **Q. Is using an interface for Participant a good design decision here, or would an abstract class offer better code reuse in this case?**

What do you think the differences are?
Is there a reason why you and your pair chose interface?

Comment (pkeugine) [inline src/main/kotlin/blackjack/model/Participant.kt]: **Q. Does the Participants class help improve clarity and structure, or could it be simplified further?**

I think it does help improve the clarity and structure.
However, I'm curious about that further improvement. What is it like?

File: src/main/kotlin/blackjack/controller/Game.kt (hunk 9 above)
Comment (pkeugine) [reply src/main/kotlin/blackjack/controller/Game.kt]: **Q. Any suggestions for improving testability?**

I think some comments that I've written will actually help improve the testability.
Also, the comment about adding some more cases about cards in hand will also improve the effectiveness as well 👍

**Review by pkeugine — CHANGES_REQUESTED** [review]
Hello Farhana! 🙌 
Nice work on blackjack mission, it was easy to read and follow 👍 

I've added some comments so be sure to check them out.
There are questions I included that I think will help you maintain this project.
Once you're done, hit me with another review request.

If you have any questions, feel free to leave comments here or ask via DM.
Happy hacking! 🎲 🎲 🎲

File: src/main/kotlin/blackjack/view/InputView.kt (hunk 1 above)
Comment (farhanaahmed) [reply src/main/kotlin/blackjack/view/InputView.kt]: Thanks for your nice suggestion.
I think class is used to create multiple instances with different states. One must instantiate it manually. An object is created as a single instance automatically and used for utilities or shared logic.
Since I’m using `InputView` and `OutputView` without needing multiple instances, defining them as `object` is better.

File: src/main/kotlin/blackjack/model/Suit.kt (hunk 4 above)
Comment (farhanaahmed) [reply src/main/kotlin/blackjack/model/Suit.kt]: I completely agree. Placing visual effects and card display inside `OutputView` makes a lot of sense since it's tied to the console output strategy.
That way, if we ever change the UI (like switching to a web interface or GUI), we can easily replace `OutputView` to another implementation without touching the game logic. It keeps our architecture flexible. I have removed visual logics from enum classes and added a method in `OutputView` to show these symbols.

File: src/main/kotlin/blackjack/model/Dealer.kt (hunk 5 above)
Comment (farhanaahmed) [reply src/main/kotlin/blackjack/model/Dealer.kt]: Yes, you are right. My pair and I wanted to finish the program first and then refactor the magic numbers later. But in the end, we didn’t have enough time to refactor our code.In the new code, I have assigned a name to it and declare it as a companion object.

File: src/test/kotlin/blackjack/model/PlayerTest.kt (hunk 6 above)
Comment (farhanaahmed) [reply src/test/kotlin/blackjack/model/PlayerTest.kt]: You're absolutely right. Using shared fixtures can lead to tightly coupled tests and unpredictable behavior. 😅

Right now, the tests are passing because:
Each test method creates and operates on the same `Player` reference, but none of them modify or depend on shared state across tests (e.g., no static/global object or reused instance between tests).

But it can fail if:
A test modifies the shared `Player`’s state (like adding cards) and another test assumes a clean slate.

To avoid these risks and make the tests more robust, I have updated the tests so each one creates its own new `Player` instance. That way, tests are completely isolated and safe from side effects.

Thanks for bringing this up, very helpful! 🙏

File: src/main/kotlin/blackjack/controller/Game.kt (hunk 7 above)
Comment (farhanaahmed) [reply src/main/kotlin/blackjack/controller/Game.kt]: You're right. Both `drawCard()` and `updateActiveStatus()` are tightly coupled with the player's own behavior. Wrapping them in a single method like `drawAndUpdate(deck)` improves encapsulation and makes the calling code cleaner. 🙌
I'll go ahead and refactor this into a method on `Player`. Thanks for the tip. :D

File: src/main/kotlin/blackjack/model/Participant.kt (hunk 8 above)
Comment (farhanaahmed) [reply src/main/kotlin/blackjack/model/Participant.kt]: Thanks for this insight. I agree that having a mutable `isActive` property can introduce the risks of accidental state changes and bugs.

Since we can determine whether a participant is active by evaluating their current total, it makes sense to remove the `isActive` variable and instead provide a function like `isStillInGame()` that calculates this.

This approach improves immutability and reduces the chance of inconsistent states. I have changed the code and added a function.

File: src/main/kotlin/blackjack/controller/Game.kt (hunk 9 above)
Comment (farhanaahmed) [reply src/main/kotlin/blackjack/controller/Game.kt]: Thanks for the suggestion. While moving the comparison logic into the `Player` class could improve encapsulation, in my current implementation I decided to keep the result evaluation inside `OutputView` for simplicity and centralized display logic.
This keeps the `Player` class focused only on maintaining its state, while `OutputView` handles both displaying cards and determining win/loss based on final totals.
I’ve also added extra test cases in `PlayerTest` to validate edge cases, especially around tricky Ace values and totals near 21, to ensure the scoring logic remains accurate

File: src/main/kotlin/blackjack/model/Deck.kt (hunk 10 above)
Comment (farhanaahmed) [reply src/main/kotlin/blackjack/model/Deck.kt]: Thanks for the feedback! I updated the giveCard() method to draw(count: Int = 1): List<Card>, allowing multiple cards to be drawn at once. This makes the deck API more flexible and helps simplify the game logic by reducing repetitive calls.I also added validation to ensure that deck has enough cards before drawing.

File: src/main/kotlin/blackjack/controller/Game.kt (hunk 11 above)
Comment (farhanaahmed) [reply src/main/kotlin/blackjack/controller/Game.kt]: Thanks for the suggestion. Making functions `private` when they’re not used externally is a great practice to encapsulate implementation details and reduce unintended usage.

I have reviewed the entire project and marked all such internal helper methods `private` to keep the API clean and focused.I hope that I didn't miss any.

File: src/main/kotlin/blackjack/view/OutputView.kt (hunk 12 above)
Comment (farhanaahmed) [reply src/main/kotlin/blackjack/view/OutputView.kt]: Thanks for the great insights. I haven’t formally defined a method ordering convention yet, but I agree that consistency helps maintainability and readability across the project.

I’m thinking of adopting this general order for all classes:

1. Public methods first (the main API).
2. Then protected/internal methods.
3. Private helper methods at the bottom.
4. Group related methods together by functionality.
5. Keep lifecycle or entry-point methods near the top.

I’ll try to follow this convention throughout the project for better structure and easier navigation.

Would you recommend any other ordering rules or best practices you’ve found useful?

File: src/test/kotlin/blackjack/model/CardTest.kt (hunk 13 above)
Comment (farhanaahmed) [reply src/test/kotlin/blackjack/model/CardTest.kt]: Thanks for the insightful suggestion! I agree that using the enum values directly in assertions is clearer and less fragile than relying on `toString()`. I have updated the tests to compare enums directly.

File: src/test/kotlin/blackjack/model/DeckTest.kt (hunk 14 above)
Comment (farhanaahmed) [reply src/test/kotlin/blackjack/model/DeckTest.kt]: I have added a `countCards()` method to `Deck` and updated tests to use it instead of accessing the internal card list directly. 
Thanks for the suggestion! 🙌

File: src/test/kotlin/blackjack/model/PlayerTest.kt (hunk 15 above)
Comment (farhanaahmed) [reply src/test/kotlin/blackjack/model/PlayerTest.kt]: Yes, it would be great to test with different cases. I have added more unit tests for this class with different values.

File: src/main/kotlin/blackjack/model/Participant.kt (hunk 16 above)
Comment (farhanaahmed) [reply src/main/kotlin/blackjack/model/Participant.kt]: Good question! 😊
Initially, we extracted `checkAces` because it was a simple, reusable line, and we wanted to keep the main logic clean.
But I can see now the main `totalValueOfCards` logic is actually *more* complex and would benefit even more from being in its own method. I didn’t realize it at first. Fixed it now, and the code looks cleaner. Thanks.

File: src/main/kotlin/blackjack/view/InputView.kt (hunk 17 above)
Comment (farhanaahmed) [reply src/main/kotlin/blackjack/view/InputView.kt]: Haha 😄 good point. I agree that being a bit more forgiving would make the experience smoother for the player.
I have updated the logic to keep prompting the user until they enter a valid input instead of throwing an exception right away. Thanks for the suggestion! 🙌

File: src/main/kotlin/blackjack/model/Participant.kt (hunk 18 above)
Comment (farhanaahmed) [reply src/main/kotlin/blackjack/model/Participant.kt]: Great question. Both interface and abstract class can work here, but as far as I understood, there are some key differences : 

- **Interface**:
    - Defines a contract without state.
    - Allows classes to implement multiple interfaces (multiple inheritance of type).
    - Usually better for flexibility and multiple behaviors mixed in.
- **Abstract class**:
    - Can hold common state and concrete method implementations.
    - Enables code reuse by providing shared functionality in one place.
    - Only allows single inheritance.
  
In our case, since `Participant` has common properties (like `cardsInHand`) and shared methods (like calculating totals), an **abstract class** might actually be better to avoid duplicating code.
We chose an interface initially to keep things flexible and simple, but after seeing the shared logic, switching to an abstract class would improve code reuse and maintainability. I have updated my code and made `Participant` an abstract class. Please let me know your thoughts about this change. 🙏

Comment (farhanaahmed) [reply src/main/kotlin/blackjack/model/Participant.kt]: Right now, `Participant` is doing *many* things:

- It stores all players and the dealer
- It handles dealing cards
- It controls player turns and dealer turns

But I had a thought to give some of its responsibilities to a different class :

- ✅ Keeping `Participant` only for storing and organizing players
- ✅ Moving turn-handling logic to another class like `ManageTurn`

But I am not sure if it’s a better idea or not. Looking forward to your valuable insights on this. 🙏

File: src/main/kotlin/blackjack/model/Suit.kt (hunk 4 above)
Comment (pkeugine) [reply src/main/kotlin/blackjack/model/Suit.kt]: PERFECTO EXPLANATION-O with examples such as GUI and web interface 💯

File: src/main/kotlin/blackjack/controller/Game.kt (hunk 9 above)
Comment (pkeugine) [reply src/main/kotlin/blackjack/controller/Game.kt]: Nice work on adding in extra test cases 💯 

Also, while I do respect and follow reviewee's opinion if their backing reasons are pretty good for most of the times, I think determining who wins and comparing players' & dealer's cards in OutputView is not a good strategy.

The reason is because just like we've talked about, `OutputView` that we're currently using is for console printing. Even if we change it to other output view strategy, we would still need that logic where we compare players' & dealer's cards and determine who wins.

However, I do understand that your implementation keeps `Player` class focus on maintaining state. Well, let's think about it while working on the next step :)

File: src/main/kotlin/blackjack/view/OutputView.kt (hunk 12 above)
Comment (pkeugine) [reply src/main/kotlin/blackjack/view/OutputView.kt]: You're method actually seems perfect to me 😂

File: src/main/kotlin/blackjack/model/Participant.kt (hunk 18 above)
Comment (pkeugine) [reply src/main/kotlin/blackjack/model/Participant.kt]: It's great 👍 
This is also a common way of using an abstract class:
1. by making features first, 
2. then finding out what the shared logics are
3. and refactor using abstract class

Nice job!

Comment (pkeugine) [reply src/main/kotlin/blackjack/model/Participant.kt]: I think it's actually fine as is. Having `ManageTurn` class would make it more difficult to manage,
since it really feels like we're trying to separate the logic because we `feel like` a class has too much responsibility.
But to me, since `Participants` has the list of players and dealers, it has the responsibility to get the deck and deal cards to them.

This also applies to [your previous comment about `Player` having too much responsibility](https://github.com/herotechcourse/kotlin-blackjack/pull/12#discussion_r2203583781). When we say `single responsibility` in SRE, I don't think it really means one unit of responsibility, such as.. `having a collection`, `dealing cards`, etc. When the class has all the ingredients such as `list of players` and `a dealer` in it, all the things related to them would be it's single responsibility.

If you have any other thoughts, I'm always here 😆

**Review by pkeugine — APPROVED** [review]
Hello again, Farhana! 🙌
It really feels like I've left a lot of comments, but nice work going through them all.
I've added some replies and additional thoughts on our topics, so it'll be nice to check them out as you work on your next step.

If you have any unanswered questions or thoughts, DM me anytime.
And as always,
Happy Hacking 🚀 🚀 🚀
